docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,77 @@
+
+1. Compute the list of blocks by the query time range
+2. Compute the minimum list of store-gateway instances containing the required blocks (using the information from the ring)
+3. Fetch series from ingesters and the matching store-gateway instances
+4. Merge and deduplicate received series
+   - Optimization: can be skipped if the querier hits only 1 store-gateway
+
+### Query results consistency
+
+When a querier executes a query, it should guarantee that either all blocks matching the time range are queried or the query fails.
+
+However, due to the (intentional) lack of a strong coordination between queriers and store-gateways, and the ring topology which can change any time, there's no guarantee that the blocks assigned to a store-gateway shard are effectively loaded on the store-gateway itself at any given point in time.
+
+The idea is introduce a **consistency check in the querier**. When a store-gateway receives a request from the querier, the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself. The querier can then merge the list of block IDs received from all store-gateway hit, and match it against the list of block IDs computed at the beginning of the query execution.
+
+There are three possible scenarios:
+
+1. The list match: all good
+2. All the blocks known by the querier are within the list of blocks returned by store-gateway, but the store-gateway also included blocks unknown to the querier: all good (it means the store-gateways have discovered and loaded new blocks before the querier discovered them)
+3. Some blocks known by the querier are **not** within the list of blocks returned by store-gateway: potential consistency issue
+
+We want to protect from a partial results response which may occur in the case #3. However there's one legit case, which would lead to frequent false positives, we want to handle: the querier has discovered new blocks before the store-gateway successfully discovered and loaded them. To protect from this case, we exclude from the consistency check the blocks which have been uploaded in the last X time (same technique already used in other Thanos components). This value X should be greater than the `-experimental.tsdb.bucket-store.consistency-delay`, because we do expect the querier to consider a block for consistency check once it's reasonably safe to assume that its store-gateway already loaded it.
+
+## Trade-offs
+
+The proposed solution comes with the following trade-offs:
+
+- A querier is not ready until it has completed an initial full scan of the bucket, downloading the `meta.json` file of every block
+- A store-gateway is not ready until it has completed an initial full scan of the bucket, downloading the `meta.json` and index header of each block matching its shard
+- If a querier hits 2+ store-gateways it may receive duplicated series if the 2+ store-gateways share some blocks due to the replication factor
```

Comment:
All of these look fine to me. Querier is already fetching meta.json and index headers, so it will do less work now.

Querier will also do deduplication, as it does today when handling results from multiple streaming ingesters.